### PR TITLE
Improve the APIs in the LDAP package

### LINKED ISSUE #6567: Implement LDAP Connector
**Description:**
$subject

### PR DESCRIPTION
**Pull request by Nuvindu**

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#
```
Comment (LakshanWeerasinghe) [inline .gitattributes:1]: Replace this file with
```
# Ensure all Java files use LF.
*.java eol=lf
```

File: build-config/resources/Ballerina.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+distribution = "2201.8.6"
+
+[platform.java17]
+graalvmCompatible = true
+
+[[platform.java17.dependency]]
+groupId = "io.ballerina.lib"
+artifactId = "ldap-native"
+version = "0.1.0"
+path = "../native/build/libs/ldap-native-0.1.0-SNAPSHOT.jar"
```
Comment (LakshanWeerasinghe) [inline build-config/resources/Ballerina.toml]:
```suggestion
path = "../native/build/libs/ldap-native-@project.version@.jar"
```

File: native/src/main/java/io/ballerina/lib/ldap/Ldap.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+import static io.ballerina.lib.ldap.Utils.ENTRY_NOT_FOUND;
+
+/**
+ * This class handles APIs of the LDAP client.
+ */
+public class Ldap {
+    private Ldap() {
+    }
+
+    public static BError initLdapConnection(BObject ldapClient, BMap<BString, Object> config) {
```
Comment (LakshanWeerasinghe) [inline native/src/main/java/io/ballerina/lib/ldap/Ldap.java:68]:
```suggestion
    public static Object initLdapConnection(BObject ldapClient, BMap<BString, Object> config) {
```

File: build-config/resources/Ballerina.toml (hunk 2 above)
Comment (Nuvindu) [reply build-config/resources/Ballerina.toml]: Fixed in [31c3a85](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/31c3a85552371b3d20a942de6e0a653f601d91cd)

File: .gitattributes (hunk 1 above)
Comment (Nuvindu) [reply .gitattributes:1]: Fixed in [2a31fb6](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/2a31fb6b1523983ad8d5dd04b3e57f10cd71427e)

File: native/src/main/java/io/ballerina/lib/ldap/Ldap.java (hunk 3 above)
Comment (Nuvindu) [reply native/src/main/java/io/ballerina/lib/ldap/Ldap.java:68]: Decided to keep the `BError` type as it also can be used here as the return type

File: ballerina/client.bal [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    remote isolated function add(string distinguishedName, record {|anydata...;|} entry)
+        returns LDAPResponse|Error = @java:Method {
+        'class: "io.ballerina.lib.ldap.Ldap"
+    } external;
+
+    # Removes an entry in a directory server.
+    #
+    # + distinguishedName - The distinguished name of the entry to remove
+    # + return - `error` if the operation fails or `()` if successfully updated
+    remote isolated function delete(string distinguishedName) returns LDAPResponse|Error = @java:Method {
```
Comment (Bhashinee) [inline ballerina/client.bal]: Shouldn't this `LDAPResponse` be `LdapResponse`?

File: ballerina/types.bal [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+# + telephoneNumber - telephone number of the person
+public type Person record {
+    string|string[]|ObjectClass|ObjectClass[] objectClass;
+    string sn;
+    string cn;
+    string userPassword?;
+    string telephoneNumber?;
+};
+
+enum ObjectClass {
```
Comment (Bhashinee) [inline ballerina/types.bal]: Shouldn't this be a public enum that is visible to users?

File: ballerina/error.bal [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+# Represents any error related to Ballerina LDAP module
+public type Error distinct error<ErrorDetails>;
+
+# The error details type for the Ballerina LDAP module.
+#
+# + resultStatus - The status of the error
+# + message - The error message
+public type ErrorDetails record {|
+    string resultStatus?;
```
Comment (Bhashinee) [inline ballerina/error.bal]:
```suggestion
    string resultCode?;
```
I think `result code` is a more familiar phrase in LDAP. WDYT?

File: ballerina/tests/test.bal [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        hostName: "111.111.11.111",
+        port: port,
+        domainName: domainName,
+        password: password
+    });
+    test:assertTrue(ldapClient is Error);
+}
+
+@test:Config {}
+public function testInvalidDomainInClient() returns error? {
```
Comment (Bhashinee) [inline ballerina/tests/test.bal]:
```suggestion
public function testInvalidDomainInClient() {
```

File: ballerina/client.bal (hunk 4 above)
Comment (Nuvindu) [reply ballerina/client.bal]: Fixed in [37aaf59](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/37aaf59ce89d002eb5bd0b08a8586dbbb2bd0665)

Comment (Nuvindu) [reply ballerina/client.bal]: Fixed in [cdde708](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/cdde70852f32464ef384105c5366885fa9486d98)

File: ballerina/types.bal (hunk 5 above)
Comment (Nuvindu) [reply ballerina/types.bal]: Fixed in [cdde708](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/cdde70852f32464ef384105c5366885fa9486d98)

File: ballerina/error.bal (hunk 6 above)
Comment (Nuvindu) [reply ballerina/error.bal]: Fixed in [cdde708](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/cdde70852f32464ef384105c5366885fa9486d98)

File: ballerina/tests/test.bal (hunk 7 above)
Comment (Nuvindu) [reply ballerina/tests/test.bal]: Fixed in [cdde708](https://github.com/ballerina-platform/module-ballerina-ldap/pull/45/commits/cdde70852f32464ef384105c5366885fa9486d98)

**Review by Bhashinee — APPROVED** [review]

**Pull request merged**